src/net/connection_controller.rs

```diff
@@ -0,0 +1,158 @@
+};
+use crossbeam_channel::Sender;
+use log::error;
+use std::{self, net::SocketAddr, time::Instant};
+
+/// Controls all aspects of the connection:
+/// * Processes incoming data (from a socket) or events (from a user).
+/// * Updates connection state: resends dropped packets, sends heartbeat packet, etc.
+/// * Creates new connections.
+/// * Checks if connection should be dropped.
+/// It doesn't own connections, but only owns necessary components to process them.
+#[derive(Debug)]
+pub struct ConnectionController<PacketSender> {
+    config: Config,
+    packet_sender: PacketSender,
+    event_sender: Sender<SocketEvent>,
+}
+
+/// Defines a connection type.
+type Connection = VirtualConnection;
+/// Defines a user event type.
+type UserEvent = Packet;
+/// Defines a connection event type.
+type ConnectionEvent = SocketEvent;
+
+impl<PacketSender: SocketSender> ConnectionController<PacketSender> {
+    /// Creates a new instance of `ConnectionHandler`.
+    pub fn new(
+        config: Config,
+        packet_sender: PacketSender,
+        event_sender: Sender<ConnectionEvent>,
+    ) -> Self {
+        ConnectionController {
+            config,
+            packet_sender,
+            event_sender,
+        }
+    }
+
+    /// Creates new connection. Also will init it and send connection event to a user.
+    pub fn create_connection(
+        &self,
+        address: SocketAddr,
+        time: Instant,
+        initial_data: Option<&[u8]>,
+    ) -> Connection {
+        // emit connect event if this is initiated by remote host.
+        if initial_data.is_some() {
+            self.event_sender
+                .send(ConnectionEvent::Connect(address))
+                .unwrap();
+        }
+        Connection::new(address, &self.config, time)
+    }
+
+    /// Determine if this connection should be dropped due to its state.
+    pub fn should_drop(&self, connection: &mut Connection, time: Instant) -> bool {
+        let should_drop = connection.packets_in_flight() > self.config.max_packets_in_flight
+            || connection.last_heard(time) >= self.config.idle_connection_timeout;
+        if should_drop {
+            self.event_sender
+                .send(ConnectionEvent::Timeout(connection.remote_address))
+                .unwrap();
+        }
+        should_drop
+    }
+
+    /// Handle a packet received from a socket.
+    pub fn handle_packet(&mut self, connection: &mut Connection, payload: &[u8], time: Instant) {
+        match connection.process_incoming(payload, time) {
+            Ok(packets) => {
+                for incoming in packets {
+                    self.event_sender
+                        .send(ConnectionEvent::Packet(incoming.0))
+                        .unwrap();
+                }
+            }
+            Err(err) => error!("Error occured processing incomming packet: {:?}", err),
+        }
+    }
+
+    /// Handle an event received from a user.

```

Comment:
Can we make more specific comments, I am not sure what we can understand under 'Handle' an event.

Reply:
This is not a public interface, I'm not sure if it requires a lot of details... and there is not many of them
Would "Process a received event and send a packet." be better?